### DISCUSSION TITLE
covariant type annotation of BaseModel

### DISCUSSION THREAD

**Discussion by ben-marengo-msmg** [pydantic/pydantic#3459] [Question] [answered]
my high level aim is to parse many json files with the correct pydantic model based on their filename, then output all the failures as a `list[ValidationError]`. 

i would like a method as follows that takes a generic model and deserialises it, returning either the error or the data

```python
from typing import Type, Any, Union

from pydantic import BaseModel, ValidationError


class M1(BaseModel):
    a: int


class M2(BaseModel):
    b: str


def parse_as_error_or_model(model: Type[BaseModel], d: dict[str, Any]) -> Union[ValidationError, BaseModel]:
    try:
        return model(**d)
    except ValidationError as ve:
        return ve


maybe_m1: M1 = parse_as_error_or_model(M1, {"a": 123})
maybe_m2: M2 = parse_as_error_or_model(M2, {"b": "bbb"})
```

mypy doesn't like it because it hasn't worked out that BaseModel should be covariant
```bash
$ mypy pydantic_test.py
pydantic_test.py:21: error: Incompatible types in assignment (expression has type "Union[ValidationError, BaseModel]", variable has type "M1")
pydantic_test.py:22: error: Incompatible types in assignment (expression has type "Union[ValidationError, BaseModel]", variable has type "M2")
```

is there a way to properly type annotate my `parse_as_error_or_model` function so that mypy knows that the type of model passed in will be the model returned?

(i already messed around a lot with typing Generic and TypeVar without any success!)

**Comment by jasujm** [comment, marked as answer]
You mentioned you've already tried `TypeVar` without success, but how did it fail exactly? You could use bound `TypeVar`s if that would help:

```
T = TypeVar("T", bound=BaseModel)

def parse_as_error_or_model(model: Type[T], d: Dict[str, Any]) -> Union[ValidationError, T]:
    try:
        return model(**d)
    except ValidationError as ve:
        return ve
```

But also `mypy` cannot possibly know that e.g. `parse_as_error_or_model(M1, {"a": 123})` returns `M1` and not `ValidationError`, so you either need explicit cast or annotating `maybe_m1` as `Union[M1, ValidationError]` -- depending on your use case.

```
maybe_m1: M1 = cast(M1, parse_as_error_or_model(M1, {"a": 123}))
# OR
maybe_m1: Union[M1, ValidationError] = parse_as_error_or_model(M1, {"a": 123})
```

**Reply by ben-marengo-msmg** [reply]
facepalm. im sure this is one of the first things i tried but somehow must not have?! works thanks!
